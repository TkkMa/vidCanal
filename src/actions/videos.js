import database from '../firebase/firebase';

export const setSearchKey = ({
    text='',
    reRender=true,
    didMount=true
})=>({
    type:'SET_SEARCH_KEY',
    query: {text, reRender, didMount}
});

export const setDidMount = ({
    didMount= true
}) => ({
    type: 'SET_DID_MOUNT',
    didMount
});

export const setVideos = ({
        searchKey = '',
        updatedHits = [],
        nextPageToken = ''    
    } = {}
) => ({
    type: 'SET_VIDEOS',
    videos: {searchKey, updatedHits, nextPageToken}
});

export const loadViewedVideos =()=>{
    return(dispatch)=>{
        return database.ref('history')
                .once('value')
                .then((snapshot)=>{
                    const videos = [];

                    snapshot.forEach((childSnapshot)=>{
                        videos.push({
                            DB_id: childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    });
                    if(videos.length){
                        dispatch(selectVideo({
                            searchKey: videos[videos.length-1].searchKey,
                            uniqueVideos:videos,
                            didMount: true
                        }));
                    }
                })

    }
}

//-- Sets the first video from list & user clicked video in Redux Store
export const selectVideo = (video)=>({
    type: 'SELECT_VIDEO',
    video
});

//-- will dispatch selectVideo by returning a function and save visited videos
export const startSelectVideo = (videoData={})=>{
    return (dispatch)=>{
        const {
            searchKey = '',
            updatedHitSelect = [],
            video = [],
            uniqueVideos = [],
            viewedAt = '',
            isSaved = false,
            reRender = true,
            didMount= true
        } = videoData;
        const videoObj = {searchKey, updatedHitSelect, video, uniqueVideos, viewedAt, isSaved, reRender, didMount};
 
        return database.ref('history').push({...video[0], viewedAt, searchKey, isSaved}).then((ref)=>{
            videoObj.uniqueVideos=[{...video[0], DB_id:ref.key, viewedAt, searchKey, isSaved}];
            dispatch(selectVideo(videoObj));
        })
    };
};

export const saveVideo = (updatedVideos)=>({
    type: 'SAVE_VIDEO',
    updatedVideos
});

// Do a few things
//--(1) modify the isSaved field in history database
//--(2) modify visitedVideos array in redux
//--(3) Save videos in saved folder in database
export const startSaveVideo = (videoData={})=>{
    return (dispatch) =>{
        const {
            isSaved= false,
            updatedVideos=[],
            dbIds=[]
        } = videoData;

        const promiseArray=[];
        dbIds.forEach((id)=>{
            promiseArray.push(database.ref(`history/${id}/isSaved`).set(isSaved))  
        });

        Promise.all(promiseArray).then(()=>{
            dispatch(saveVideo(updatedVideos));
        })
    }
}

export const clearVideoHistory =()=>({
    type: 'CLEAR_VIDEO_HISTORY'
})

export const startClearVideoHistory = ()=>{
    return(dispatch) =>{
        return database.ref('history').remove().then(()=>{
            console.log("Remove succeeded");
            dispatch(clearVideoHistory());
        })
    }
}