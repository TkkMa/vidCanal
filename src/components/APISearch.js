const axios = require('axios');
const moment = require('moment');

const ROOT_URL = {
    YT: 'https://www.googleapis.com/youtube/v3/search',
    D: 'https://api.dailymotion.com/videos',
    V: 'https://api.vimeo.com/videos'
}
const ROOT_URL_VIDEO = {
    YT: 'https://www.googleapis.com/youtube/v3/videos'
    // D: 'https://api.dailymotion.com/video/',
    // V: 'https://api.vimeo.com/videos'
};
const API_KEY= {
    YT:'AIzaSyDWU_q-QSxsSUECDdeWI8WBnVE2q07p804',
    V: 'f0ca0fbcd4918f82208f732dfded4341'
}

const vidAPISearch = (options={}, engine='YT')=>{
    let params;
    if (engine === 'YT'){
        params = {
            maxResults: options.num,
            part: 'snippet',
            key: API_KEY.YT,
            q: options.term,
            type: 'video',
            order: options.sortByVal,
            pageToken: options.pageToken[engine]
        };
        if(options.sortTimeVal){params.publishedAfter=options.sortTimeVal;}
        return axios.get(ROOT_URL[engine], { params: params })
            .then(response=>{           
                return {
                    pageInfo: response.data.pageInfo, 
                    items: response.data.items, 
                    nextPageToken: response.data.nextPageToken || ''
                };
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    } else if(engine==='D'){
        params = {
            limit: options.num,
            search: options.term,
            sort: options.sortByVal,
            page: options.pageToken[engine],
            fields: 'id,title,owner.screenname,owner.url,views_total,embed_url,created_time,description,thumbnail_60_url,thumbnail_120_url,thumbnail_180_url'
          };

        if(options.sortTimeVal){params.created_after=moment(options.sortTimeVal).unix();}
        if(!params.sort){console.error('No alphabetical sorting exists for Dailymotion'); return null}
        return axios.get(ROOT_URL[engine], {params: params})
            .then(response=>{
                const arrayOfObj = response.data.list.map(({
                    ['owner.screenname']: ownerScreenname,
                    ['owner.url']: ownerUrl,
                    ...rest
                })=>({
                    ownerScreenname,
                    ownerUrl,
                    ...rest
                }))
                return {
                    pageInfo: {
                        total: response.data.total,
                        has_more: response.data.has_more
                    }, 
                    items: arrayOfObj, 
                    nextPageToken: options.pageToken[engine]+1
                };            
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    } else if(engine==='V'){
        params = {
            access_token: API_KEY.V,
            per_page: options.num,
            query: options.term,
            page: options.pageToken[engine],
            sort: options.sortByVal,
            fields:'total,page,per_page,paging,data,uri,pictures,name,release_time,user,stats,description,embed'
        }

        // if(options.sortTimeVal){params.filter=moment(options.sortTimeVal).unix();}
        return axios.get(ROOT_URL[engine], { params: params })
            .then(response=>{
                return {
                    pageInfo: {
                        paging: response.data.paging,
                        total: response.data.total
                    }, 
                    items: response.data.data, 
                    nextPageToken: options.pageToken[engine]+1
                };
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    }  
}

//-- Daily motion and Vimeo only requires one API Call in the previous search request to get all fields
//-- Youtube requires two API calls
const indVidAPISearch = (options, engine='YT') =>{
    let params;
    if(engine==='YT'){
        params={
            id: options.id,
            key: API_KEY.YT,
            part: 'snippet, statistics'
        }

        return axios.get(ROOT_URL_VIDEO[engine], {params: params})
            .then(response=>{
            return response.data.items
            })
            .catch(function(error) {
            console.error(error);
        });
    } else if(engine==='D'){
        return [options];
    } else if(engine==='V'){
        return [options];
    }
};

module.exports = {
    vidAPISearch,
    indVidAPISearch
}