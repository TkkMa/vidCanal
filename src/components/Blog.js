import React from 'react';
import { Link } from 'react-router-dom';

const Blog = () => (
  <div className="construction">
    <img src='/images/under-construction-image.png' />
    <Link to="/">Go home</Link>
  </div>
);

export default Blog;