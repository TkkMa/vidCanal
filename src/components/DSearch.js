const axios = require('axios');
const ROOT_URL = 'https://api.dailymotion.com/videos';
// const ROOT_URL_VIDEO = 'https://www.googleapis.com/youtube/v3/videos';
// const API_KEY='AIzaSyDWU_q-QSxsSUECDdeWI8WBnVE2q07p804';

const DSearch = (options={})=>{

  const params = {
    limit: options.num,
    search: options.term,
    sort: options.sortBy,
    page: options.pageToken
  };
  if(options.sortTimeVal){params.createdAfter=options.sortTimeVal;}

  return axios.get(ROOT_URL, { params: params })
    .then((response)=>{
      return {
        has_more: response.data.has_more,
        items: response.data.list, 
        nextPageToken: response.data.page + 1
      };
    })
    .catch(function(error) {
      console.error(error);
    });  
}

const YTVideo = (options) =>{
  const params={
    id: options.id,
    key: API_KEY,
    part: 'snippet, statistics'
  }

  return axios.get(ROOT_URL_VIDEO, {params: params})
    .then((response)=>{
      return response.data.items
    })
    .catch(function(error) {
      console.error(error);
    });
}

module.exports = {
  YTSearch,
  YTVideo
}