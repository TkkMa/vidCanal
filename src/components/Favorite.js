import React from 'react';
import FavoriteListFilters from './FavoriteListFilters';
import FavoriteList from './FavoriteList';

const Favorite = ()=>{
    return(
        <div>
            <FavoriteListFilters />
            <FavoriteList />
        </div>
    )
}

export default Favorite