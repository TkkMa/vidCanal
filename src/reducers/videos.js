const DEFAULT_QUERY = 'redux';

const videosReducerDefaultState = {
    searchKey: DEFAULT_QUERY,
    results: {
       YT: null,
       D: null,
       V: null 
    },
    selectedVideo: {},
    visitedVideos:[],
    resultDetail: {
        YT: null,
        D: null,
        V: null 
    },
    reRender: true,
    didMount: true
};

export default (state = videosReducerDefaultState, action) =>{
    switch (action.type){
        case 'SET_SEARCH_KEY':
            return{
                ...state,
                searchKey: action.query.text,
                reRender: action.query.reRender,
                didMount: action.query.didMount
            };
        case 'SET_DID_MOUNT':
            return{
                ...state,
                didMount: action.didMount
            }
        case 'SET_VIDEOS':
            return {
                ...state,
                results: {
                    ...state.results,
                    [action.videos.engine]: {
                        ...state.results[action.videos.engine],
                        [action.videos.searchKey]: {hits:action.videos.updatedHits}
                    }
                }
            };
        case 'SELECT_VIDEO':
            return{
                ...state,
                didMount: action.video.didMount,
                searchKey: action.video.searchKey,
                resultDetail:{
                    ...state.resultDetail,
                    [action.video.engine]:{
                        ...state.resultDetail[action.video.engine],
                        [action.video.searchKey]:{hits:action.video.updatedHitSelect}
                    }    
                },
                selectedVideo: action.video.uniqueVideos[0],
                visitedVideos: [...state.visitedVideos, ...action.video.uniqueVideos],
                reRender: action.video.reRender
            };
        case 'SAVE_VIDEO':
            return{
                ...state,
                visitedVideos: action.updatedVideos
            };
        case 'CLEAR_VIDEO_HISTORY':
            return{
                ...state,
                visitedVideos: []
            };
        case 'SET_RE_RENDER':
            return{
                ...state,
                reRender: action.reRender
            }
        default:
            return state;
    }
};