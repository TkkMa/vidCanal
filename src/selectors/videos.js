import moment from 'moment';
// Get videos on display after filtering...etc

export const videoHistory = (videoArray, {startDate, endDate, text, isSaved}) =>{
    return videoArray.filter((video)=>{
            const viewedAtMoment = moment(video.viewedAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(viewedAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(viewedAtMoment,'day') : true;
            const textMatch = video.searchKey.toLowerCase().includes(text.toLowerCase()) || 
                                video.snippet.title.toLowerCase().includes(text.toLowerCase());
            const savedMatch = (isSaved) ? video.isSaved : true;
            return startDateMatch && endDateMatch && textMatch && savedMatch;
        }).sort((a, b) => {
            return moment(b.viewedAt).diff(moment(a.viewedAt));
        });
}

export const vidSearchInputs =(pageToggle, uploadDate, stateValues) => new Promise((resolve)=>{
    let {pageActive, maxViewedPage} = stateValues;
    console.log('pageActive', pageActive);
    switch (pageToggle){
        case 'first_page':
            pageActive = 1;
            break;
        case 'chevron_left':
            pageActive = Math.max(pageActive-1, 1);
            break;
        case 'chevron_right':
            pageActive = pageActive + 1;
            maxViewedPage = Math.max(pageActive, maxViewedPage);
            break;
        case 'last_page':
            pageActive = maxViewedPage;
            break;
        default:
            pageActive = 1;
            maxViewedPage = 1;
    }
    console.log('pageActive after', pageActive);
    let sortTimeVal;
    switch(uploadDate){
        case 'today':
            sortTimeVal = moment().startOf('day').format();
            break;
        case 'week':
            sortTimeVal = moment().startOf('week').format();
            break;
        case 'month':
            sortTimeVal =  moment().startOf('month').format();
            break;
        case 'year':
            sortTimeVal = moment().startOf('year').format();
            break;
        case 'fiveYears':
            sortTimeVal = moment().subtract(5, 'years').format();
            break;
        case 'allTime':
            sortTimeVal = '';
            break;                               
        default:
            sortTimeVal = '';
    };
    resolve({pageActive, maxViewedPage, sortTimeVal});
}) 