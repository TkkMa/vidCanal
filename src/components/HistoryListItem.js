import React, {Component} from "react";
import {connect} from 'react-redux';
import moment from "moment";
import {startSelectVideo, startSaveVideo} from "../actions/videos";
import {addFavCount, removeFavCount} from '../actions/filters';

export class HistoryListItem extends Component{

    state = {
        height: "150px"
    }

    onVideoSelect = () =>{
        const {video, resultDetail} = this.props;
        console.log(`clicked video: ${video}`);
        console.log(`Existing hits in resultDetail: ${resultDetail}`);
        this.props.startSelectVideo({
            searchKey: video.searchKey,
            updatedHitSelect: (resultDetail[video.searchKey] && 
                                resultDetail[video.searchKey].hits) || [],
            video: [video],
            viewedAt: moment().utc().toISOString(),
            isSaved: video.isSaved,
            reRender: false,
            didMount: false
        });

        this.props.history.push('/');
    }

    updateIsSavedStatus= (result) =>{
        const dbIds = [];
        const updatedVideos = this.props.visitedVideos.map(video =>{
            if(video.id===this.props.video.id){
                video.isSaved = result.isSaved;
                dbIds.push(video.DB_id);
            }
            return video;
        });
        this.props.startSaveVideo({
            isSaved: result.isSaved,
            updatedVideos,
            dbIds
        });
    }

    onVideoSave=(e)=>{
        e.stopPropagation();
        let isSavedStatus = false;
        if($(e.target).text() === "star_border"){
            $(e.target).text("star");
            isSavedStatus = true;
            this.props.addFavCount({
                count: this.props.count + 1,
                videoId: this.props.video.id,
                ids: this.props.ids
            });
        } else{
            $(e.target).text("star_border");
            const index = this.props.ids.findIndex(idElement=> idElement.videoId===this.props.video.id);
            const count = (index>-1) ? this.props.count - 1 : this.props.count;
            if (index > -1){ 
                this.props.removeFavCount({
                    count,
                    videoIds: this.props.ids.filter(idElement => idElement.videoId !== this.props.video.id),
                    foundDbId: this.props.ids[index].DB_id
                })
            };
        }
        this.updateIsSavedStatus({isSaved: isSavedStatus});
    }

    render(){
        const {video} = this.props;
        const imageUrl = video.snippet && video.snippet.thumbnails.medium.url || video.DB_id;
        // console.log(imageUrl);
        return(
            (!video.snippet)? (
                <li className="collection-item avatar">
                    <div className="row">
                        <div className="col s12">
                            MISSING VIDEO
                        </div>
                    </div>
                </li>
            ) : (
                <li className="HLI-1 collection-item avatar" onClick={()=>this.onVideoSelect()}>
                    <div className="row">
                        <div className="col s2 div-record-1">
                            <span className="span-search-text">Search term: </span>
                                {video.searchKey}
                            <span className="span-search-text">Last Visited Time:</span>
                            <span>{moment(video.viewedAt).local().format('ddd, MMM D YYYY')}</span>
                            <span>{moment(video.viewedAt).local().format('h:mm:ss a')}</span>                    
                        </div>
                        <div className="col s3 div-record-2">
                            <img src={imageUrl} className="responsive-img"/>
                        </div>
                        <div className="col s6 div-record-3">
                            <span className="title">{video.snippet.title}</span>
                            <p>{moment(video.snippet.publishedAt).format('DD MMM YYYY')} - {video.statistics.viewCount} views<br />
                                Uploaded by:<a href={`http://www.youtube.com/channel/${video.snippet.channelId}`}> {video.snippet.channelTitle}</a>
                            </p>
                            <p>{video.snippet.description.substring(0,200)}...</p>                    
                        </div>
                        <div className="col s1 div-record-4">
                            <i className="material-icons" onClick={this.onVideoSave}>{(video.isSaved)?'star' : 'star_border'}</i> 
                        </div>
                    </div>
                </li>
            )
        )
    }
}

const mapStateToProps = (state)=>({
    visitedVideos: state.videos.visitedVideos,
    count: state.filters.unViewedFavCount,
    ids: state.filters.unViewedFavIds,
});

const mapDispatchToProps = (dispatch)=>({
    startSelectVideo: (video)=> dispatch(startSelectVideo(video)),
    startSaveVideo: (video)=> dispatch(startSaveVideo(video)),
    addFavCount: (favCount) => dispatch(addFavCount(favCount)),
    removeFavCount: (favCount) => dispatch(removeFavCount(favCount))
});

export default connect(mapStateToProps, mapDispatchToProps)(HistoryListItem);