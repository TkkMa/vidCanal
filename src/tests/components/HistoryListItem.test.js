import React from 'react';
import {shallow} from 'enzyme';
import {HistoryListItem} from '../../components/HistoryListItem';
import videos from '../fixtures/videos';
import moment from 'moment';

test('should render History video list item', ()=>{
    const wrapper = shallow(<HistoryListItem 
                                key={videos[0].id+1}
                                video={videos[0]}
                            />);
    expect(wrapper).toMatchSnapshot();
})

test('should render no video list items with message', ()=>{
    const wrapper = shallow(<HistoryListItem 
                                video={{}}
                            />);
    expect(wrapper).toMatchSnapshot();
})

test('should revert to / after item click', ()=>{
    const startSelectVideoSpy = jest.fn();
    const history = {push: jest.fn()};
    const resultDetail={
        'london tourism': {
            hits: []
        }
    };
    const wrapper = shallow(<HistoryListItem
                                key={videos[0].id+1}
                                video={videos[0]}
                                startSelectVideo={startSelectVideoSpy}
                                history={history}
                                resultDetail={resultDetail}
                            />);
    wrapper.find('li').at(0).simulate('click');
    expect(startSelectVideoSpy).toHaveBeenLastCalledWith({
        searchKey: videos[0].searchKey,
        updatedHitSelect: [],
        video: [videos[0]],
        viewedAt: moment().utc().toISOString(),
        isSaved: videos[0].isSaved,
        reRender:false,
        didMount: false
    });
    expect(history.push).toHaveBeenLastCalledWith('/');
})