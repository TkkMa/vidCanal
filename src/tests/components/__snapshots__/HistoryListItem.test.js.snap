// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render History video list item 1`] = `
<li
  className="HLI-1 collection-item avatar"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col s2 div-record-1"
    >
      <span
        className="span-search-text"
      >
        Search term: 
      </span>
      london tourism
      <span
        className="span-search-text"
      >
        Last Visited Time:
      </span>
      <span>
        Thu, Dec 25 1969
      </span>
      <span>
        8:00:00 am
      </span>
    </div>
    <div
      className="col s3 div-record-2"
    >
      <img
        className="responsive-img"
        src="https://i.ytimg.com/vi/45ETZ1xvHS0/mqdefault.jpg"
      />
    </div>
    <div
      className="col s6 div-record-3"
    >
      <span
        className="title"
      >
        London Vacation Travel Guide | Expedia
      </span>
      <p>
        07 Mar 2013
         - 
        2849731
         views
        <br />
        Uploaded by:
        <a
          href="http://www.youtube.com/channel/UCGaOvAFinZ7BCN_FDmw74fQ"
        >
           
          Expedia
        </a>
      </p>
      <p>
        http://www.expedia.com/London.d178279.Destination-Travel-Guides

A trip to London might as well be a trip through history. This city has long been a hub for finance, education, and culture, and is ful
        ...
      </p>
    </div>
    <div
      className="col s1 div-record-4"
    >
      <i
        className="material-icons"
        onClick={[Function]}
      >
        star
      </i>
    </div>
  </div>
</li>
`;

exports[`should render no video list items with message 1`] = `
<li
  className="collection-item avatar"
>
  <div
    className="row"
  >
    <div
      className="col s12"
    >
      MISSING VIDEO
    </div>
  </div>
</li>
`;
